registry/mdns_registry.go

```diff
@@ -298,13 +314,24 @@ func (m *mdnsRegistry) Deregister(service *Service, opts ...DeregisterOption) er
 		newEntries[0].node.Shutdown()
 		delete(m.services, service.Name)
 	} else {
-		m.services[service.Name] = newEntries
+		m.services[options.Domain][service.Name] = newEntries
 	}
 
 	return nil
 }
 
 func (m *mdnsRegistry) GetService(service string, opts ...GetOption) ([]*Service, error) {
+	// parse the options, fallback to the default domain
+	var options GetOptions
+	for _, o := range opts {
+		o(&options)
+	}
+	if len(options.Domain) == 0 {
+		m.Lock()
```

Comment:
At some point we should revisit how we're using the locks. Here you're locking to read from `m.opts` but elsewhere you use the same lock to lock access to `m.services`. It's thread safe, but could be made more efficient with specific locks for whatever you're trying to lock.

Reply:
You're correct. I need to revisit how this has been implemented. I think options we shouldn't  need a lock on as they shouldn't be mutated whilst the service is running.